### ISSUE TITLE
override def trigger = always, for Auto Plugin

### ISSUE THREAD

**Issue by eed3si9n** [sbt/sbt#2093]
This will bring in the requirements as if it's been enabled by hand.

``` scala
object FooPlugin extends AutoPlugin {
  override def requires = com.typesafe.sbt.GitVersioning
  override def trigger = always
}
```

**Comment by jsuereth** [comment]
I'm against this feature.  I realize there are use cases made hard by the fact that one plugin author can make his plugin be something you have to manually specify.  I think we should honor those wishes.

What this allows is someone to come along and auto-enable a plugin that was not intended to be auto-enabled.  THEN when the bug reports show up on the original plugin, the author has a pissing match with whomever owns the `always` plugin.   I'd rather have the inverted discussion, where the dependent plugin comes to the original author and asks for the plugin to be auto-enabled and the original one is _designed_ for that scenario to begin with.

What we're doing here is asking for internet fights.  At least, that's my logic behind not liking this feature.

P.S.  I can work on a way to make the GitVersioning plugin be auto-enabled :)

**Comment by eed3si9n** [comment]
Making `GitVersioning` an autoplugin doesn't really solve the issue: Some _plugins_ should remain optional, but some _organizations_ want them auto-enabled. Note that I am not suggesting change for GitVersioning; however, I am suggesting the auto-enablement of `FooPlugin`.

This is semantically equivalent to going to every single subproject and adding `enablePlugins(FooPlugin)`, which will bring in its required plugins. When we flipped the logic to `requires` we've already paved the way for this change. I don't think it would result to internet fights just because a plugin is automatically pulled in by _another_ plugin.

**Comment by paulp** [comment]
I came to report this. It was specifically GitVersioning I wanted to depend on, not that it matters.

Look at the choice you're presenting to a plugin author with the existing trigger options.

1) I can depend on no other plugins, and my plugin will be added to projects automatically with a single line. Installation instructions are to add one line to one file and it just works.
2) If I have _any_ dependencies, it is _mandatory_ that every project manually call enablePlugins with either my plugin or with all of my dependencies. Installation instructions explode in complexity and number of possible variations.

This line you are drawing is completely arbitrary. If there is some value to not having to manually enable a plugin - and obviously there is - why make it impossible to have dependencies? Dependencies don't have anything to do with anything. I can already put any code I want in the plugin, what are you protecting someone from?

I'm going to import all the source of sbt-git into my plugin so it can still be automatically triggered, that's the workaround. Doesn't seem like an improvement from a holistic standpoint. You can't reduce complexity or insecurity by making things inconvenient. You can only spread it around.

**Comment by dwijnand** [comment]
(By the by, here's an alternative workaround: https://github.com/sbt/sbt-houserules/blob/v0.3.1/src/main/scala/sbt/houserules/BetterGitVersionPlugin.scala)

**Comment by paulp** [comment]
@dwijnand Are you saying that will work correctly even if the project doesn't manually enable the GitVersioning plugin? Because that has to be a bug in sbt! If I can magically make the project depend on another plugin by adding a build setting which uses it, then it would be truly pointless to prevent me from depending on it more straightforwardly.

And if the project has to enable GitVersioning for that to work, then nothing has been achieved.

**Comment by dwijnand** [comment]
Yeah it doesn't require `GitVersioning` to be enabled - it's effectively re-defining `GitVersioning`, but now it auto-triggers.

**Comment by paulp** [comment]
You're bringing in the git plugin code as a "classic" dependency rather than a "plugin" dependency. What happens if they do enable GitVersioning? What happens if they explicitly include a slightly different version in their plugins.sbt, is there some facility for resolving conflicts which arise this way?

Mostly it seems to reinforce what I said earlier, that people are going to reuse code and you can't stop them. The more you push dependencies outside of the usual mechanisms, the more every project's dependency logic becomes a special unique snowflake.

**Comment by paulp** [comment]
I think I'm still confused how your workaround works. By what means does GitPlugin become enabled in the project, given that they have not listed it as a dependency?

**Comment by paulp** [comment]
@dwijnand The method you describe DOES work for GitPlugin but DOES NOT work for GitVersioning. I see now, GitPlugin is an auto-triggering autoplugin which doesn't add any settings; GitVersioning is a manually-triggered autoplugin which depends on GitPlugin and adds settings, the same ones you added. So GitVersioning is useless to me when we're working around this way.

Thanks for the idea.

**Comment by dwijnand** [comment]
[Started writing this before the last comment..]

I'm not 100% sure what you mean by "they have not listed it as a dependency ", but GitPlugin is configured to auto-triggers (as opposed to GitVersioning) and is defined as a requirement for BetterGitVersionPlugin, and sbt-git is a dependency of sbt-houserules.

And yeah I'm not sure what happens if `GitVersioning` is also enabled. For the same version (or versions where the functionality is effectively the same) it's just duplicate work done during build loading. If GitVersioning does something different in the future, then it might break.

A mildly more robust solution would be one in which all properties of GitVersioning are "forwarded" in the "better" version, eg. in case GitVersioning gains some globalSettings or projectSettings. This could be generalised, but it would further undermine the counter-argument @jsuereth is making.

**Comment by dwijnand** [comment]
Also, credit to @eed3si9n for the idea (https://github.com/sbt/sbt-houserules/commit/70d335369dabe4dd0220d23244fff8e2e38623a0), I'm just the idea messenger.

**Comment by eed3si9n** [comment]
### opt-in/opt-out problem

The original design of `sbt.Plugin` automatically brought in every settings defined under `projectSettings` to all subprojects. This was considered too aggressive (because not all subprojects wants all the plugins) and setting loading order was arbitrary, so you couldn't depend on other plugins reliably. For example you have `jsEngine := X.JavaScript` and `jsEngine := Y.TypeScript`, you won't know what `jsEngine` you get.

### settings dependency and loading order problem

The plugin authors worked around this by never using `projectSettings`, and declaring something like [`assemblySettings`](https://github.com/sbt/sbt-assembly/tree/0.11.2), which each build users needed to add to all subprojects. This enabled per-subproject opt-in of plugins, and explicit setting loading order, but it pushed the onus of knowing the exact order to each build user, which was bad. For example you have `jsEngine := X.JavaScript` and `jsEngine := Y.TypeScript`, wrong order could pick wrong JavaScript engine.

### `sbt.AutoPlugin`

`sbt.AutoPlugin` solves the opt-in/opt-out problem by creating enabled/weak-enabled/disabled state. A plugin that's been brought in by triggering mechanism can be disabled via `disablePlugins(...)`.

`sbt.AutoPlugin` also solves the setting loading order problem by topologically sorting the settings based on the declared plugin requirements.

The choice of current triggers `allRequirements` and `noTrigger` stems from trying to avoid logical inconsistency, I think. Let's say plugin `C` requires `A` and `B`, and it's triggered via `AllRequirements`. Enabling `A` and disabling `B` won't load `C`.
- Suppose we do introduce `always`, this will weakly enable `A`, `B`, and `C`.
- Next if you disable `B`, would it cause all `A`, `B`, and `C` to be disabled?

**Comment by paulp** [comment]
@eed3si9n [So familiar...](https://www.wikiwand.com/en/Multiple_inheritance#/The_diamond_problem)

**Comment by dwijnand** [comment]
Thinking of the use cases behind the change proposed here, what about being able to enable or disable plugins build-wide?  Then you would be able to:

```scala
buildWideEnablePlugins(GitVersioning)
buildWideDisablePlugins(BintrayPlugin)
```

**Comment by eed3si9n** [comment]
I wonder if we could reuse the existing DSL entry as follows?

```scala
ThisBuild / enablePlugins(GitVersioning)
ThisBuild / disablePlugins(BintrayPlugin)
```

**Comment by dwijnand** [comment]
That would be introducing scoping to `DslEntry`... I'm not sure.

We need some "mixfix operators" support :P https://twitter.com/dwijnand/status/1181131852872126464

**Comment by eed3si9n** [comment]
I am closing this due to inactivity.
